src/thread.rs

```diff
@@ -0,0 +1,583 @@
+//! Implements threads.
+
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+
+use log::trace;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::DefId;
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::{
+    middle::codegen_fn_attrs::CodegenFnAttrFlags,
+    mir,
+    ty::{self, Instance},
+};
+
+use crate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SchedulingAction {
+    /// Execute step on the active thread.
+    ExecuteStep,
+    /// Execute destructors of the active thread.
+    ExecuteDtors,
+    /// Stop the program.
+    Stop,
+}
+
+/// A thread identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ThreadId(usize);
+
+/// The main thread. When it terminates, the whole application terminates.
+const MAIN_THREAD: ThreadId = ThreadId(0);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(usize::try_from(id).unwrap())
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(usize::try_from(id).unwrap())
```

Comment:
`From` impls should never panic, that is a key part of their contract.

Reply:
Since it seems like our concurrency primitives assume all thread IDs fit in 32bit, what about making `ThreadId` wrap `u32` instead of `usize`?